File: TypeScript/tests/two_sum.test.ts
import { assertEquals } from "@std/assert";
import { twoSum, twoSum2 } from "../src/two_sum.ts";

interface TestCase {
  nums: number[];
  target: number;
  want: number[];
}

const testCases: TestCase[] = [
  { nums: [2, 7, 11, 15], target: 9, want: [0, 1] },
  { nums: [3, 2, 4], target: 6, want: [1, 2] },
  { nums: [3, 3], target: 6, want: [0, 1] },
  { nums: [2, 3, 4, 1, 25, 8], target: 30, want: [-1] },
];

testCases.forEach((testCase, index) => {
  Deno.test(`twoSum - test case ${index + 1}`, () => {
    const result = twoSum(testCase.nums, testCase.target);
    assertEquals(result, testCase.want);
  });

  Deno.test(`twoSum2 - test case ${index + 1}`, () => {
    const result = twoSum2(testCase.nums, testCase.target);
    assertEquals(result, testCase.want);
  });
});

File: C++/tests/two_sum_test.cpp
#include <gtest/gtest.h>
#include <two_sum.hpp>

class TwoSumTest : public ::testing::Test {
protected:
  Solution solution;
};

TEST_F(TwoSumTest, increasing_order) {
  vector<int> nums = {2, 7, 11, 15};
  int target = 9;
  vector<int> want = {0, 1};

  vector<int> actual = solution.twoSum(nums, target);
  EXPECT_EQ(actual, want);
}

TEST_F(TwoSumTest, different_order) {
  vector<int> nums = {3, 2, 4};
  int target = 6;
  vector<int> want = {1, 2};

  vector<int> actual = solution.twoSum(nums, target);
  EXPECT_EQ(actual, want);
}

TEST_F(TwoSumTest, same_numbers) {
  vector<int> nums = {3, 3};
  int target = 6;
  vector<int> want = {0, 1};

  vector<int> actual = solution.twoSum(nums, target);
  EXPECT_EQ(actual, want);
}

TEST_F(TwoSumTest, no_solution) {
  vector<int> nums = {1, 2, 3};
  int target = 10;
  vector<int> want = {};

  vector<int> actual = solution.twoSum(nums, target);
  EXPECT_EQ(actual, want);
}

TEST_F(TwoSumTest, empty_vector) {
  vector<int> nums = {};
  int target = 10;
  vector<int> want = {};

  vector<int> actual = solution.twoSum(nums, target);
  EXPECT_EQ(actual, want);
}
